nncf/quantization/algorithms/accuracy_control/evaluator.py

```diff
@@ -0,0 +1,172 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Iterable, List, Optional, Tuple, TypeVar, Union
+
+from nncf.common.factory import EngineFactory
+from nncf.data.dataset import Dataset
+from nncf.quantization.algorithms.accuracy_control.backend import AccuracyControlAlgoBackend
+
+TModel = TypeVar("TModel")
+
+
+class Output:
+    """
+    Contains ordered list of raw model outputs.
+    """
+
+    def __init__(self):
+        self._outputs = []
+
+    def __iter__(self):
+        return iter(self._outputs)
+
+    def register(self, value: Any) -> None:
+        """
+        Registers `value` as a raw model output.
+        """
+        self._outputs.append(value)
+
+
+class Evaluator:
+    """
+    Evaluator encapsulates a logic to validate model and collect values for each item.
+    The value is either calculated metric or model output. This is determined by the
+    `Evaluator.is_metric_mode()` method.
+    """
+
+    def __init__(
+        self,
+        validation_fn: Callable[[Any, Iterable[Any]], Tuple[float, Union[List[float], List[Output], None]]],
+        algo_backend: AccuracyControlAlgoBackend,
+    ):
+        """
+        :param validation_fn: Validation function to validate model.
+        :param algo_backend: The `AccuracyControlAlgoBackend` algo backend.
+        """
+        self._validation_fn = validation_fn
+        self._algo_backend = algo_backend
+        self._metric_mode = None
+
+    def is_metric_mode(self) -> bool:
+        """
+        Returns mode of `Evaluator`.
+
+        :return: A boolean indicator where `True` means that the `Evaluator` collects
+            metric value for each item and `False` means that the `Evaluator` collects
+            logits for each item.
+        """
+        return self._metric_mode
+
+    def validate(
+        self, model: TModel, dataset: Dataset, indices: Optional[List[int]] = None
+    ) -> Tuple[float, Union[List[float], List[Output], None]]:
+        """
+        Validates model.
+
+        :param model: Model to validate.
+        :param dataset: Dataset to validate the model.
+        :param indices: Zero-based indices of data items that should be selected from
+            the dataset.
+        :return: A tuple (metric_value, values_for_each_item) where
+            - metric_values: This is a metric for the model.
+            - values_for_each_item: If the `Evaluator.is_metric_mode()` condition is true,
+                then `values_for_each_item` represents the list of metric value for each item.
+                Otherwise, if the condition is false, it represents list of logits for each
+                item.
+        """
+        model_for_inference = self._algo_backend.prepare_for_inference(model)
+        if self._metric_mode is None:
+            self._determine_mode(model_for_inference, dataset)
+
+        if not self.is_metric_mode() and indices is not None:
```

Comment:
Could you briefly explain why `validate` method does not support indices for is_metric_mode == False?

Reply:
We can not validate (i.e. validation_fn only works for the full dataset) the model on the subset when `is_metric_mode == False`.